expr/interval_compound_literal.go

```diff
@@ -0,0 +1,246 @@
+}
+
+func (m IntervalCompoundLiteral) ToProtoLiteral() *proto.Expression_Literal {
+	t := m.getType()
+	intrCompPB := &proto.Expression_Literal_IntervalCompound{}
+
+	if m.years != 0 || m.months != 0 {
+		yearToMonthProto := &proto.Expression_Literal_IntervalYearToMonth{
+			Years:  m.years,
+			Months: m.months,
+		}
+		intrCompPB.IntervalYearToMonth = yearToMonthProto
+	}
+
+	if m.days != 0 || m.seconds != 0 || m.subSeconds != 0 {
+		dayToSecondProto := &proto.Expression_Literal_IntervalDayToSecond{
+			Days:          m.days,
+			Seconds:       m.seconds,
+			PrecisionMode: &proto.Expression_Literal_IntervalDayToSecond_Precision{Precision: m.precision.ToProtoVal()},
+			Subseconds:    m.subSeconds,
+		}
+		intrCompPB.IntervalDayToSecond = dayToSecondProto
+	}
+
+	return &proto.Expression_Literal{
+		LiteralType:            &proto.Expression_Literal_IntervalCompound_{IntervalCompound: intrCompPB},
+		Nullable:               t.GetNullability() == types.NullabilityNullable,
+		TypeVariationReference: t.GetTypeVariationReference(),
+	}
+}
+
+func (m IntervalCompoundLiteral) ToProto() *proto.Expression {
+	return &proto.Expression{RexType: &proto.Expression_Literal_{
+		Literal: m.ToProtoLiteral(),
+	}}
+}
+
+func intervalCompoundLiteralFromProto(l *proto.Expression_Literal) Literal {
+	icLiteral := IntervalCompoundLiteral{}.WithNullability(getNullability(l.Nullable))
+	yearToMonth := l.GetIntervalCompound().GetIntervalYearToMonth()
+	if yearToMonth != nil {
+		icLiteral = icLiteral.WithYears(yearToMonth.Years).WithMonths(yearToMonth.Months)
+	}
+	dayToSecond := l.GetIntervalCompound().GetIntervalDayToSecond()
+	if dayToSecond == nil {
+		// no day to second part
+		return icLiteral
+	}
+	err := validateIntervalDayToSecondProto(dayToSecond)
+	if err != nil {
+		return nil
+	}
+
+	// get subSecond/precision value from proto. To get value it takes care of deprecated microseconds
+	precision, subSeconds, err := intervalCompoundPrecisionSubSecondsFromProto(dayToSecond)
+	if err != nil {
+		return nil
+	}
+	return icLiteral.WithDays(dayToSecond.Days).WithSeconds(dayToSecond.Seconds).WithSubSecond(subSeconds, precision)
+}
+
+func (IntervalCompoundLiteral) isRootRef()            {}
+func (m IntervalCompoundLiteral) GetType() types.Type { return m.getType() }
+func (m IntervalCompoundLiteral) String() string {
+	return fmt.Sprintf("%s(years:%d,months:%d, days:%d, seconds:%d subseconds:%d)",
+		m.getType(), m.years, m.months, m.days, m.seconds, m.subSeconds)
+}
+func (m IntervalCompoundLiteral) Equals(rhs Expression) bool {
+	if other, ok := rhs.(IntervalCompoundLiteral); ok {
+		return m.getType().Equals(other.GetType()) && (m == other)
+	}
+	return false
+}
+
+func (m IntervalCompoundLiteral) ToProtoFuncArg() *proto.FunctionArgument {
+	return &proto.FunctionArgument{
+		ArgType: &proto.FunctionArgument_Value{Value: m.ToProto()},
+	}
+}
+
+func (m IntervalCompoundLiteral) Visit(VisitFunc) Expression { return m }
```

Comment:
shouldn't we return `visit(m)` ?

Reply:
I kept is similar to other literal implementations (e.g. [here](https://github.com/substrait-io/substrait-go/blob/cbd28cb19499af1923484ec82540350528249075/expr/literals.go#L501))